File: solutions/2026-07-29/java/2026-07-29-smallest-palindromic-rearrangement-ii.java
```java
// Finds the k-th lexicographically smallest palindromic permutation of a given palindromic string.
// Link: https://leetcode.com/problems/smallest-palindromic-rearrangement-ii/
//
// Approach:
// 1. Count the frequency of each character in the input string `s`.
// 2. For a string to be rearranged into a palindrome, at most one character can have an odd frequency.
//    Since the input `s` is guaranteed to be a palindrome, this condition is already met.
// 3. To form a palindromic rearrangement, we only need to arrange the first half of the palindrome.
//    The second half will be the reverse of the first half. If the string length is odd, the middle character
//    is fixed by the character with the odd count.
// 4. We count the characters that will form the first half. For each character with frequency `f`,
//    `f/2` instances will go into the first half.
// 5. The number of distinct palindromic permutations is equivalent to the number of distinct permutations
//    of the characters in the first half. This can be calculated using the multinomial coefficient formula:
//    N! / (n1! * n2! * ... * nk!), where N is the total number of characters in the first half, and
//    n_i is the count of the i-th distinct character in the first half.
// 6. We need to calculate the k-th lexicographically smallest permutation of the first half.
//    This is a standard algorithm:
//    a. For each position in the first half (from left to right), iterate through the available characters
//       in lexicographical order.
//    b. For each character, calculate how many permutations start with that character. This is done by
//       considering the remaining characters and their counts, and calculating the number of permutations
//       of those remaining characters.
//    c. If `k` is less than or equal to this count, then the current character is the one for the current
//       position. Append it to the result, decrement its count, and move to the next position.
//    d. If `k` is greater than this count, subtract this count from `k` and try the next character.
// 7. Pre-compute factorials and inverse factorials (or use `BigInteger` for combinations) to efficiently
//    calculate the number of permutations.
// 8. If the total number of distinct palindromic permutations is less than `k`, return an empty string.
//
// Time Complexity: O(L + A*L), where L is the length of `s` and A is the size of the alphabet (26).
//                  Counting frequencies takes O(L). Calculating total permutations can take O(A log(L!)).
//                  The k-th permutation generation takes O(A * L) in the worst case for each digit,
//                  but since we iterate through available characters and calculate combinations, it's more
//                  accurately O(A * L) overall because combination calculation with precomputed factorials is O(1).
//                  However, if we don't precompute or use BigInteger directly, it can be slower. With precomputation,
//                  it's closer to O(L) for frequency counting and O(A*L) for permutation construction.
//                  Since A is constant (26), the complexity is dominated by O(L).
// Space Complexity: O(A), where A is the size of the alphabet, for storing character counts and factorials.
class Solution {
    // Using BigInteger for combinations to handle large numbers.
    // Max length of s is 10^4, so length of first half is 5000. Factorials can be very large.

    // Precompute factorials up to 5000 (max length of first half).
    private static final int MAX_HALF_LEN = 5000;
    private BigInteger[] fact = new BigInteger[MAX_HALF_LEN + 1];

    // Initialize factorials.
    private void precomputeFactorials() {
        fact[0] = BigInteger.ONE;
        for (int i = 1; i <= MAX_HALF_LEN; i++) {
            fact[i] = fact[i - 1].multiply(BigInteger.valueOf(i));
        }
    }

    // Calculates combinations C(n, k) = n! / (k! * (n-k)!).
    private BigInteger combinations(int n, int k) {
        if (k < 0 || k > n) {
            return BigInteger.ZERO;
        }
        // C(n, k) = n! / (k! * (n-k)!)
        // n! is fact[n]
        // k! is fact[k]
        // (n-k)! is fact[n-k]
        return fact[n].divide(fact[k].multiply(fact[n - k]));
    }

    // Calculates the number of distinct permutations of a multiset.
    // Given character counts, total number of characters is n.
    // The formula is n! / (count1! * count2! * ...).
    private BigInteger countPermutations(int n, int[] counts) {
        BigInteger res = fact[n]; // Start with n!
        for (int count : counts) {
            if (count > 0) {
                res = res.divide(fact[count]); // Divide by count! for each character
            }
        }
        return res;
    }

    public String smallestBeautifulString(String s, int k) {
        // Precompute factorials needed for combination calculations.
        precomputeFactorials();

        // Character counts for the entire string.
        int[] totalCounts = new int[26];
        for (char c : s.toCharArray()) {
            totalCounts[c - 'a']++;
        }

        // Character counts for the first half of the palindrome.
        // If a character has count `f`, `f/2` instances go to the first half.
        int[] halfCounts = new int[26];
        int halfLen = 0;
        char oddChar = ' '; // Character with odd frequency, if any.

        for (int i = 0; i < 26; i++) {
            halfCounts[i] = totalCounts[i] / 2;
            halfLen += halfCounts[i]; // Total length of the first half.
            if (totalCounts[i] % 2 != 0) {
                oddChar = (char) ('a' + i);
            }
        }

        // If k is larger than the total number of distinct palindromic permutations, return empty string.
        // Total palindromic permutations = permutations of the first half.
        BigInteger totalPalindromicPermutations = countPermutations(halfLen, halfCounts);
        if (BigInteger.valueOf(k).compareTo(totalPalindromicPermutations) > 0) {
            return "";
        }

        // Build the k-th lexicographically smallest permutation of the first half.
        StringBuilder firstHalf = new StringBuilder();
        int remainingLen = halfLen; // Number of characters still needed for the first half.
        BigInteger kBig = BigInteger.valueOf(k); // Use BigInteger for k comparisons.

        // Iterate through each position of the first half.
        for (int i = 0; i < halfLen; i++) {
            // Try each character from 'a' to 'z' for the current position.
            for (int charCode = 0; charCode < 26; charCode++) {
                // If this character is available in the first half.
                if (halfCounts[charCode] > 0) {
                    // Tentatively place this character at the current position.
                    halfCounts[charCode]--; // Use one instance of this character.
                    remainingLen--; // One less character needed overall.

                    // Calculate the number of permutations possible with the remaining characters.
                    // This is the number of permutations of the remaining multiset.
                    BigInteger permutationsWithRemaining = countPermutations(remainingLen, halfCounts);

                    // If k is less than or equal to the number of permutations with the remaining characters,
                    // it means the k-th permutation starts with the current character.
                    if (kBig.compareTo(permutationsWithRemaining) <= 0) {
                        firstHalf.append((char) ('a' + charCode)); // Append the chosen character.
                        // k remains the same relative to this block of permutations.
                        break; // Move to the next position.
                    } else {
                        // If k is larger, it means the k-th permutation does not start with this character.
                        // Subtract the count of permutations starting with this character from k.
                        kBig = kBig.subtract(permutationsWithRemaining);
                        // Backtrack: restore the character count and remaining length.
                        halfCounts[charCode]++;
                        remainingLen++;
                    }
                }
            }
        }

        // Construct the full palindrome.
        StringBuilder result = new StringBuilder(firstHalf);
        if (oddChar != ' ') {
            result.append(oddChar); // Append the middle character if string length is odd.
        }
        // Append the reverse of the first half.
        result.append(new StringBuilder(firstHalf).reverse());

        return result.toString();
    }
}
```

File: solutions/2026-07-29/javascript/2026-07-29-smallest-palindromic-rearrangement-ii.js
// Problem Summary: Find the k-th lexicographically smallest palindromic permutation of a given palindromic string.
// Link: https://leetcode.com/problems/smallest-palindromic-rearrangement-ii/
// Approach:
// 1. Count the frequency of each character in the input string `s`.
// 2. Since `s` is palindromic, at most one character can have an odd frequency. This character will be the middle character of any palindromic rearrangement.
// 3. Construct the first half of the palindromic string by taking half the count of each character. This first half will determine the lexicographical order.
// 4. Calculate the number of distinct palindromic permutations. This is equivalent to the number of distinct permutations of the first half string. The formula for permutations with repetitions is n! / (c1! * c2! * ...), where n is the length of the first half string and ci is the count of each character.
// 5. If `k` is greater than the total number of distinct palindromic permutations, return "".
// 6. To find the k-th permutation, we can use a technique similar to finding the k-th permutation of a regular string. We iterate through the available characters for the first half, calculate how many permutations start with that character, and subtract that count from `k` until `k` is small enough to pinpoint the character for the current position.
//    - For each position in the first half, iterate through sorted available characters.
//    - For each character, calculate the number of permutations of the remaining characters.
//    - If `k` is less than or equal to this count, the current character is the one for this position. Append it to the result, decrease its count, and move to the next position.
//    - Otherwise, subtract the count from `k` and try the next character.
// 7. Once the first half is constructed, the full palindrome is formed by concatenating the first half, the middle character (if any), and the reverse of the first half.
//
// Time Complexity: O(N + alphabet_size * N), where N is the length of s.
//   - Counting character frequencies: O(N)
//   - Calculating total permutations: O(N) due to factorial calculations with precomputed factorials/inverse factorials or O(alphabet_size * N) if calculating on the fly.
//   - Constructing the k-th permutation: For each of the N/2 positions, we iterate through up to `alphabet_size` characters. For each character, calculating permutations of remaining characters can take O(N) naively, or O(alphabet_size) if precomputed factorials are used. Thus, O(N/2 * alphabet_size * alphabet_size) or O(N/2 * alphabet_size). With precomputation, it's closer to O(N * alphabet_size).
//   - String concatenation and reversal: O(N)
//   - Precomputing factorials and inverse factorials: O(N)
// Space Complexity: O(N) for storing character counts, precomputed factorials, and the resulting string.
const smallestPalindromicRearrangementII = (s, k) => {
    // Helper function to calculate combinations (n choose k)
    const combinations = (n, k) => {
        if (k < 0 || k > n) {
            return 0;
        }
        if (k === 0 || k === n) {
            return 1;
        }
        if (k > n / 2) {
            k = n - k;
        }
        let res = 1;
        for (let i = 1; i <= k; ++i) {
            res = res * (n - i + 1) / i;
        }
        return res;
    };

    // Helper function to calculate permutations with repetitions
    // n! / (c1! * c2! * ...)
    // This function calculates the number of permutations for the remaining characters
    const countPermutations = (counts) => {
        let n = 0;
        let denominator = 1;
        for (const char in counts) {
            n += counts[char];
            denominator *= factorial[counts[char]];
        }
        return factorial[n] / denominator;
    };

    const MAX_LEN = 10004; // Maximum possible length of s + a small buffer
    const factorial = new Array(MAX_LEN);
    const invFactorial = new Array(MAX_LEN);
    const MOD = 10**9 + 7; // Not strictly needed for this problem as results can be large, but good practice if intermediate calculations might overflow if we were asked for modulo.

    // Precompute factorials
    factorial[0] = 1;
    invFactorial[0] = 1;
    for (let i = 1; i < MAX_LEN; ++i) {
        factorial[i] = (factorial[i - 1] * i);
        // invFactorial[i] = power(factorial[i], MOD - 2); // For modular inverse if MOD was used
    }

    // Count character frequencies
    const charCounts = {};
    for (const char of s) {
        charCounts[char] = (charCounts[char] || 0) + 1;
    }

    let middleChar = '';
    const halfCounts = {};
    let halfLength = 0;

    // Construct counts for the first half and find the middle character
    const sortedChars = Object.keys(charCounts).sort();
    for (const char of sortedChars) {
        if (charCounts[char] % 2 === 1) {
            middleChar = char;
        }
        halfCounts[char] = Math.floor(charCounts[char] / 2);
        halfLength += halfCounts[char];
    }

    // Calculate total distinct palindromic permutations
    // This is the number of permutations of the first half string
    const totalPermutations = countPermutations(halfCounts);

    // If k is greater than the total number of permutations, return empty string
    if (k > totalPermutations) {
        return "";
    }

    let resultHalf = "";
    let currentK = k; // Use currentK to modify k as we build the permutation

    // Build the k-th lexicographically smallest permutation of the first half
    for (let i = 0; i < halfLength; ++i) {
        // Iterate through available characters in sorted order
        for (const char of sortedChars) {
            if (halfCounts[char] > 0) {
                // Temporarily decrement count to calculate permutations of remaining characters
                halfCounts[char]--;

                // Calculate the number of permutations that can be formed with the remaining characters
                const numPermsWithRemaining = countPermutations(halfCounts);

                // If currentK is within the range of permutations starting with this character
                if (currentK <= numPermsWithRemaining) {
                    resultHalf += char;
                    // The character is chosen, break the inner loop and move to the next position
                    break;
                } else {
                    // currentK is larger, so we skip all permutations starting with this character
                    // Subtract the count of permutations we just skipped
                    currentK -= numPermsWithRemaining;
                    // Restore the count since this character was not chosen
                    halfCounts[char]++;
                }
            }
        }
    }

    // Construct the full palindrome
    // The first half is `resultHalf`
    // The middle character is `middleChar`
    // The second half is the reverse of `resultHalf`
    const result = resultHalf + middleChar + resultHalf.split('').reverse().join('');

    return result;
};
